// @microsoft/teamsfx-core/build/component/feature/sso.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.SSO = void 0;
const tslib_1 = require("tslib");
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
require("reflect-metadata");
const typedi_1 = require("typedi");
const localizeUtils_1 = require("../../common/localizeUtils");
const utils_1 = require("../../common/utils");
const plugins_1 = require("../../plugins");
require("../connection/azureWebAppConfig");
const constants_1 = require("../constants");
const debug_1 = require("../debug");
require("../resource/azureSql");
require("../resource/identity");
const utils_2 = require("../utils");
const workflow_1 = require("../workflow");
let SSO = class SSO {
    constructor() {
        this.name = "sso";
    }
    async add(context, inputs) {
        const updates = getUpdateComponents(context, inputs);
        // generate manifest
        const aadApp = typedi_1.Container.get(constants_1.ComponentNames.AadApp);
        {
            const res = await aadApp.generateManifest(context, inputs);
            if (res.isErr())
                return teamsfx_api_1.err(res.error);
        }
        // config sso
        if (updates.aad) {
            context.projectSetting.components.push({
                name: constants_1.ComponentNames.AadApp,
                provision: true,
                deploy: true,
            });
        }
        if (updates.tab) {
            const teamsTabComponent = workflow_1.getComponent(context.projectSetting, constants_1.ComponentNames.TeamsTab);
            teamsTabComponent.sso = true;
        }
        if (updates.bot) {
            const teamsBotComponent = workflow_1.getComponent(context.projectSetting, constants_1.ComponentNames.TeamsBot);
            teamsBotComponent.sso = true;
        }
        // generate bicep
        {
            const res = await aadApp.generateBicep(context, inputs);
            if (res.isErr())
                return teamsfx_api_1.err(res.error);
            const bicepRes = await utils_2.bicepUtils.persistBiceps(inputs.projectPath, utils_1.convertToAlphanumericOnly(context.projectSetting.appName), res.value);
            if (bicepRes.isErr())
                return bicepRes;
        }
        // generate auth files
        if (inputs.stage === teamsfx_api_1.Stage.addFeature &&
            inputs[plugins_1.AzureSolutionQuestionNames.Features] !== plugins_1.TabOptionItem.id) {
            const res = await aadApp.generateAuthFiles(context, inputs, updates.tab, updates.bot);
            if (res.isErr())
                return teamsfx_api_1.err(res.error);
        }
        // update app manifest
        {
            const capabilities = [{ name: "WebApplicationInfo" }];
            const appManifest = typedi_1.Container.get(constants_1.ComponentNames.AppManifest);
            const res = await appManifest.addCapability(inputs, capabilities);
            if (res.isErr())
                return teamsfx_api_1.err(res.error);
        }
        // local debug settings
        {
            const res = await debug_1.generateLocalDebugSettings(context, inputs);
            if (res.isErr())
                return teamsfx_api_1.err(res.error);
        }
        // generate config bicep
        {
            const res = await utils_2.generateConfigBiceps(context, inputs);
            if (res.isErr())
                return teamsfx_api_1.err(res.error);
        }
        // show notification
        if (inputs.platform == teamsfx_api_1.Platform.VSCode) {
            context.userInteraction
                .showMessage("info", localizeUtils_1.getLocalizedString("core.addSso.learnMore", plugins_1.AddSsoParameters.LearnMore), false, plugins_1.AddSsoParameters.LearnMore)
                .then((result) => {
                var _a;
                const userSelected = result.isOk() ? result.value : undefined;
                if (userSelected === plugins_1.AddSsoParameters.LearnMore) {
                    (_a = context.userInteraction) === null || _a === void 0 ? void 0 : _a.openUrl(plugins_1.AddSsoParameters.LearnMoreUrl);
                    context.telemetryReporter.sendTelemetryEvent(plugins_1.SolutionTelemetryEvent.AddSsoReadme, {
                        [plugins_1.SolutionTelemetryProperty.Component]: plugins_1.SolutionTelemetryComponentName,
                    });
                }
            });
        }
        else if (inputs.platform == teamsfx_api_1.Platform.CLI) {
            await context.userInteraction.showMessage("info", localizeUtils_1.getLocalizedString("core.addSso.learnMore", plugins_1.AddSsoParameters.LearnMoreUrl), false);
        }
        return teamsfx_api_1.ok({
            func: plugins_1.AddSsoParameters.AddSso,
            capabilities: [
                ...(updates.tab ? [plugins_1.AddSsoParameters.Tab] : []),
                ...(updates.bot ? [plugins_1.AddSsoParameters.Bot] : []),
            ],
        });
    }
};
SSO = tslib_1.__decorate([
    typedi_1.Service("sso")
], SSO);
exports.SSO = SSO;
function getUpdateComponents(context, inputs) {
    if (inputs.stage === teamsfx_api_1.Stage.create) {
        return {
            tab: true,
            aad: true,
        };
    }
    let needsBot = false;
    let needsTab = false;
    const aadComponent = workflow_1.getComponent(context.projectSetting, constants_1.ComponentNames.AadApp);
    const teamsBotComponent = workflow_1.getComponent(context.projectSetting, constants_1.ComponentNames.TeamsBot);
    if (teamsBotComponent && !teamsBotComponent.sso) {
        needsBot = teamsBotComponent.hosting !== constants_1.ComponentNames.Function;
    }
    const teamsTabComponent = workflow_1.getComponent(context.projectSetting, constants_1.ComponentNames.TeamsTab);
    if (teamsTabComponent && !teamsTabComponent.sso) {
        needsTab = true;
    }
    return {
        bot: needsBot,
        tab: needsTab,
        aad: !aadComponent,
    };
}
