// @microsoft/teamsfx-core/build/component/resource/azureAppService/azureFunction.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.AzureFunctionResource = void 0;
const tslib_1 = require("tslib");
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const typedi_1 = require("typedi");
const constants_1 = require("../../constants");
const azureAppService_1 = require("./azureAppService");
let AzureFunctionResource = class AzureFunctionResource extends azureAppService_1.AzureAppService {
    constructor() {
        super(...arguments);
        this.name = "azure-function";
        this.alias = "Functions";
        this.displayName = "Azure Functions";
        this.bicepModuleName = "azureFunction";
        this.outputs = constants_1.FunctionOutputs;
        this.finalOutputKeys = ["resourceId", "endpoint"];
        this.templateContext = {
            identity: {
                resourceId: constants_1.IdentityOutputs.identityResourceId.bicepVariable,
            },
        };
    }
    async configure(context, inputs) {
        return teamsfx_api_1.ok(undefined);
    }
    async deploy(context, inputs) {
        return await super.deploy(context, inputs, true);
    }
};
AzureFunctionResource = tslib_1.__decorate([
    typedi_1.Service("azure-function")
], AzureFunctionResource);
exports.AzureFunctionResource = AzureFunctionResource;
