// @microsoft/teamsfx-core/build/plugins/solution/spfx-solution/solution.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.TeamsSPFxSolution = void 0;
const tslib_1 = require("tslib");
const typedi_1 = require("typedi");
const constants_1 = require("../../../common/constants");
const constants_2 = require("../fx-solution/v3/constants");
const addFeature_1 = require("./addFeature");
let TeamsSPFxSolution = class TeamsSPFxSolution {
    constructor() {
        this.name = constants_2.BuiltInSolutionNames.spfx;
        this.displayName = constants_1.PluginDisplayName.SpfxSolution;
        this.getQuestionsForAddFeature = addFeature_1.getQuestionsForAddFeature;
        this.addFeature = addFeature_1.addFeature;
        this.publishApplication = addFeature_1.publishApplication;
    }
};
TeamsSPFxSolution = tslib_1.__decorate([
    typedi_1.Service(constants_2.BuiltInSolutionNames.spfx)
], TeamsSPFxSolution);
exports.TeamsSPFxSolution = TeamsSPFxSolution;
