// @microsoft/teamsfx-core/build/plugins/solution/utils/questions.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.selectCapabilitiesQuestion = exports.selectMultiPluginsQuestion = exports.selectMultipleFeaturesQuestion = exports.SolutionV3QuestionNames = void 0;
const question_1 = require("../fx-solution/question");
var SolutionV3QuestionNames;
(function (SolutionV3QuestionNames) {
    SolutionV3QuestionNames["capabilities"] = "capabilities";
    SolutionV3QuestionNames["features"] = "features";
    SolutionV3QuestionNames["plugins"] = "plugins";
})(SolutionV3QuestionNames = exports.SolutionV3QuestionNames || (exports.SolutionV3QuestionNames = {}));
exports.selectMultipleFeaturesQuestion = {
    name: SolutionV3QuestionNames.features,
    title: "Select features to add",
    type: "multiSelect",
    staticOptions: [],
};
exports.selectMultiPluginsQuestion = {
    name: SolutionV3QuestionNames.plugins,
    title: "Select plugins to deploy",
    type: "multiSelect",
    staticOptions: [],
};
exports.selectCapabilitiesQuestion = {
    name: SolutionV3QuestionNames.capabilities,
    title: "Select capabilities",
    type: "multiSelect",
    staticOptions: [question_1.TabOptionItem, question_1.BotOptionItem, question_1.MessageExtensionItem],
    default: [question_1.TabOptionItem.id],
    skipSingleOption: true,
};
