// @microsoft/teamsfx-core/build/plugins/resource/keyvault/v3/index.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyVaultPluginV3 = void 0;
const tslib_1 = require("tslib");
const lib_1 = require("@feathersjs/hooks/lib");
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const path = tslib_1.__importStar(require("path"));
const typedi_1 = require("typedi");
const constants_1 = require("../../../../common/constants");
const tools_1 = require("../../../../common/tools");
const CommonErrorHandlerMW_1 = require("../../../../core/middleware/CommonErrorHandlerMW");
const folder_1 = require("../../../../folder");
const question_1 = require("../../../solution/fx-solution/question");
const constants_2 = require("../../../solution/fx-solution/v3/constants");
const constants_3 = require("../constants");
let KeyVaultPluginV3 = class KeyVaultPluginV3 {
    constructor() {
        this.name = constants_2.BuiltInFeaturePluginNames.keyVault;
        this.displayName = "Key Vault Plugin";
    }
    async generateBicep(ctx, inputs) {
        const pluginCtx = { plugins: inputs.allPluginsAfterAdd };
        const bicepTemplateDirectory = path.join(folder_1.getTemplatesFolder(), "plugins", "resource", "keyvault", "bicep");
        const provisionModuleResult = path.join(bicepTemplateDirectory, constants_3.Constants.provisionModuleTemplateFileName);
        const provisionOrchestration = await tools_1.generateBicepFromFile(path.join(bicepTemplateDirectory, constants_1.Bicep.ProvisionFileName), pluginCtx);
        const provisionModules = await tools_1.generateBicepFromFile(provisionModuleResult, pluginCtx);
        const result = {
            Provision: {
                Orchestration: provisionOrchestration,
                Modules: { keyVault: provisionModules },
            },
            Reference: {
                m365ClientSecretReference: constants_3.Constants.KeyVaultBicep.m365ClientSecretReference,
                botClientSecretReference: constants_3.Constants.KeyVaultBicep.botClientSecretReference,
            },
        };
        return teamsfx_api_1.ok([result]);
    }
    async addInstance(ctx, inputs) {
        const solutionSettings = ctx.projectSetting.solutionSettings;
        const activeResourcePlugins = solutionSettings.activeResourcePlugins;
        if (!activeResourcePlugins.includes(this.name))
            activeResourcePlugins.push(this.name);
        const azureResources = solutionSettings.azureResources;
        if (!azureResources.includes(question_1.AzureResourceKeyVault.id))
            azureResources.push(question_1.AzureResourceKeyVault.id);
        return teamsfx_api_1.ok([constants_2.BuiltInFeaturePluginNames.identity]);
    }
    async updateBicep(ctx, inputs) {
        const result = {
            Reference: {
                m365ClientSecretReference: constants_3.Constants.KeyVaultBicep.m365ClientSecretReference,
                botClientSecretReference: constants_3.Constants.KeyVaultBicep.botClientSecretReference,
            },
        };
        return teamsfx_api_1.ok([result]);
    }
};
tslib_1.__decorate([
    lib_1.hooks([CommonErrorHandlerMW_1.CommonErrorHandlerMW({ telemetry: { component: constants_2.BuiltInFeaturePluginNames.keyVault } })]),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Object, Object]),
    tslib_1.__metadata("design:returntype", Promise)
], KeyVaultPluginV3.prototype, "generateBicep", null);
tslib_1.__decorate([
    lib_1.hooks([CommonErrorHandlerMW_1.CommonErrorHandlerMW({ telemetry: { component: constants_2.BuiltInFeaturePluginNames.keyVault } })]),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Object, Object]),
    tslib_1.__metadata("design:returntype", Promise)
], KeyVaultPluginV3.prototype, "addInstance", null);
tslib_1.__decorate([
    lib_1.hooks([CommonErrorHandlerMW_1.CommonErrorHandlerMW({ telemetry: { component: constants_2.BuiltInFeaturePluginNames.keyVault } })]),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Object, Object]),
    tslib_1.__metadata("design:returntype", Promise)
], KeyVaultPluginV3.prototype, "updateBicep", null);
KeyVaultPluginV3 = tslib_1.__decorate([
    typedi_1.Service(constants_2.BuiltInFeaturePluginNames.keyVault)
], KeyVaultPluginV3);
exports.KeyVaultPluginV3 = KeyVaultPluginV3;
